### PR TITLE
Ace codegen

### PR DESCRIPTION
**Pull request by Al-Kindi-0**
## Describe your changes

Adds an `ace-codegen` crate to generate ACE circuits from Miden AIR. It introduces a `RecordingAirBuilder` to capture the AIR constraints, lowers them into a DAG for the verifier's constraints evaluation check, defines READ/EVAL input layouts, and gives tools to encode such circuits.

There is also the `processor_air_chiplet_rows` test which prints the number of rows in the ACE chiplet for evaluating the circuit for `ProcessorAir`. For the currently implemented set of constraints, which contains about 95% of the constraints, it shows that it takes `5712` ACE rows to evaluate the circuit.

There is also an integration test in `crates/lib/core/tests/crypto/circuit_evaluation.rs` that builds the `ProcessorAir` ACE circuit, encodes it, streams it via `adv_pipe`, executes `eval_circuit`, and runs `prove_and_verify`, ensuring the end‑to‑end MASM path works.

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](../CONTRIBUTING.md).
- Commits are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.
- Updated `CHANGELOG.md'

### CODE REVIEW

**Comment by adr1anh** [comment]
Only did a very high-level review, and it looks really good already. I'm trying to merge the necessary changes to the SymbolicAir to upstream plonky3 so we can reduce the amount of duplicated code we need to carry around. I'll check though if we can do without and if so I'll fast track a review for here.

**Comment by adr1anh** [comment]
I took the liberty of adding a minimal SymbolicExpression/Variable implementation so we can avoid relying on any other air crates from plonky3/uni-stark/p3-miden. It should also make it easier for us to add some optimizations to the graph

File: crates/ace-codegen/src/dag/ir.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    /// Per-column coefficient vectors (highest-degree first).
+    pub coeffs: Vec<Vec<EF>>,
+}
+
+impl<EF> PeriodicColumnData<EF> {
+    /// Number of periodic columns.
+    pub fn len(&self) -> usize {
+        self.coeffs.len()
+    }
+}
```
Comment (adr1anh) [inline crates/ace-codegen/src/dag/ir.rs:85]: I'd make max period a computed function and rename Len to `num_columns`

File: crates/ace-codegen/src/dag/lower.rs [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+use crate::{
+    AceError, layout::InputKey, quotient::build_quotient_recomposition_dag,
+    randomness::RandomnessPlan,
+};
+
+/// Lower a symbolic expression into DAG nodes using the provided layout.
+pub fn lower_expr<F, EF>(
+    expr: &SymExpr<EF>,
+    builder: &mut DagBuilder<EF>,
+    layout: &crate::layout::InputLayout,
```
Comment (adr1anh) [inline crates/ace-codegen/src/dag/lower.rs]: Nit : add an import for this

File: crates/ace-codegen/src/dag/lower.rs [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    periodic_nodes: &[NodeId],
+) -> Result<NodeId, AceError>
+where
+    F: PrimeCharacteristicRing,
+    EF: PrimeCharacteristicRing + BasedVectorSpace<F> + Copy + Eq + Hash,
```
Comment (adr1anh) [inline crates/ace-codegen/src/dag/lower.rs]: Why not use Field and ExtensionField?

File: crates/ace-codegen/src/dag/builder.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+/// compact and deterministic.
+#[derive(Debug)]
+pub struct DagBuilder<EF> {
+    nodes: Vec<NodeKind<EF>>,
+    cache: HashMap<NodeKind<EF>, NodeId>,
+}
+
+impl<EF> DagBuilder<EF>
+where
+    EF: PrimeCharacteristicRing + Copy + Eq + std::hash::Hash,
```
Comment (adr1anh) [inline crates/ace-codegen/src/dag/builder.rs]: Field should be enough

File: crates/ace-codegen/src/dag/lower.rs [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+/// Build the verifier-equivalent root expression DAG.
+///
+/// This constructs the folded constraint accumulator, divides by the vanishing
+/// polynomial, recomposes the quotient, and subtracts both sides to yield the
+/// root expression evaluated by the ACE circuit.
+pub fn build_verifier_dag<F, EF>(
+    constraints: &[SymExpr<EF>],
+    layout: &crate::layout::InputLayout,
```
Comment (adr1anh) [inline crates/ace-codegen/src/dag/lower.rs]: Same as above

File: crates/ace-codegen/src/dag/lower.rs [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+    periodic: Option<&PeriodicColumnData<EF>>,
+) -> Result<AceDag<EF>, AceError>
+where
+    F: PrimeCharacteristicRing,
+    EF: PrimeCharacteristicRing + BasedVectorSpace<F> + Copy + Eq + Hash,
```
Comment (adr1anh) [inline crates/ace-codegen/src/dag/lower.rs]: Same as above

File: crates/ace-codegen/src/dag/lower.rs [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+{
+    if layout.counts.ext_degree != EF::DIMENSION {
+        return Err(AceError::InvalidExtensionDegree {
```
Comment (adr1anh) [inline crates/ace-codegen/src/dag/lower.rs]: I don't think the parameters are super helpful, this should never happen

File: crates/ace-codegen/src/dag/lower.rs [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+    let root = builder.sub(folded, quotient);
+
+    Ok(AceDag { nodes: builder.into_nodes(), root })
+}
+
+/// Convert periodic columns into evaluation coefficients for DAG building.
+///
+/// The periodic columns are provided as evaluations; this function applies an
+/// inverse DFT so the DAG can evaluate them at `z_k` inside the circuit.
+pub fn build_periodic_data<F, EF>(
```
Comment (adr1anh) [inline crates/ace-codegen/src/dag/lower.rs]: This would make more sense as a constructor on PeriodicData

File: crates/ace-codegen/src/dag/lower.rs [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+        let coeff_row = values.into_iter().map(EF::from).collect();
+        coeffs.push(coeff_row);
+    }
+
+    Ok(PeriodicColumnData { max_len, coeffs })
+}
+
+fn build_periodic_nodes<EF>(
+    builder: &mut DagBuilder<EF>,
+    layout: &crate::layout::InputLayout,
```
Comment (adr1anh) [inline crates/ace-codegen/src/dag/lower.rs]: Same as earlier

File: crates/ace-codegen/src/randomness.rs [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        if num == 0 {
+            return (RandomnessPlan::Direct { num }, None, None);
+        }
+
+        if num_inputs == num {
+            return (RandomnessPlan::Direct { num }, None, None);
+        }
+
+        if num_inputs == 2 {
+            // Stored as [beta, alpha] in memory words.
```
Comment (huitseeker) [inline crates/ace-codegen/src/randomness.rs]: `RandomnessPlan::from_layout` chooses `Direct` whenever `num_randomness_inputs == num_randomness`. Since `input_counts_for_air` hard‑codes `num_randomness_inputs = 2`, any AIR with `num_randomness == 2` will not use the alpha/beta expansion and will map challenges to `InputKey::Randomness`(0/1) instead of `[alpha, 1, beta, ...]`. This conflicts with the docs in layout/mod.rs and this comment.

File: crates/ace-codegen/src/dag/lower.rs [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+        return Err(AceError::InvalidPeriodicColumn { index: 0, count: periodic.coeffs.len() });
+    }
+
+    let mut cache = HashMap::<usize, NodeId>::new();
+    let mut nodes = Vec::with_capacity(periodic.coeffs.len());
+    for (idx, coeffs) in periodic.coeffs.iter().enumerate() {
+        if coeffs.is_empty() {
+            nodes.push(builder.constant(EF::ZERO));
+            continue;
+        }
```
Comment (adr1anh) [inline crates/ace-codegen/src/dag/lower.rs]: Isn't this an error?

File: crates/ace-codegen/src/randomness.rs [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+                alpha_idx,
+                beta_idx,
+            );
+        }
+
+        (RandomnessPlan::Direct { num }, None, None)
+    }
+
+    /// Build the plan from a concrete input layout.
+    pub(crate) fn from_layout(layout: &InputLayout) -> Result<Self, AceError> {
```
Comment (huitseeker) [inline crates/ace-codegen/src/randomness.rs]: This returns `Direct` when `num_randomness_inputs == num_randomness` before checking the alpha/beta path. A regression test for `num_randomness == 2` seems important.

File: crates/ace-codegen/src/dag/lower.rs [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+        return Err(AceError::InvalidPeriodicColumn { index: 0, count: max_len });
+    }
+
+    let dft = Radix2DitParallel::<F>::default();
+    let mut coeffs = Vec::with_capacity(periodic_table.len());
+    for (idx, col) in periodic_table.into_iter().enumerate() {
+        if col.is_empty() {
+            coeffs.push(Vec::new());
+            continue;
+        }
```
Comment (adr1anh) [inline crates/ace-codegen/src/dag/lower.rs]: This should be an error

File: crates/ace-codegen/src/dag/lower.rs [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+        let values = dft.idft(col);
+        let coeff_row = values.into_iter().map(EF::from).collect();
+        coeffs.push(coeff_row);
+    }
+
+    Ok(PeriodicColumnData { max_len, coeffs })
+}
+
+fn build_periodic_nodes<EF>(
```
Comment (adr1anh) [inline crates/ace-codegen/src/dag/lower.rs:135]: The checks can be ignored imo, since we can assume PeriodicColumnData was correctly constructed and validated

File: crates/ace-codegen/src/dag/lower.rs [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+        let col_len = coeffs.len();
+        let max_len = periodic.max_len;
+        if !max_len.is_multiple_of(col_len) || !max_len.is_power_of_two() {
+            return Err(AceError::InvalidPeriodicColumn { index: idx, count: col_len });
+        }
+        let ratio = max_len / col_len;
+        if !ratio.is_power_of_two() {
+            return Err(AceError::InvalidPeriodicColumn { index: idx, count: col_len });
+        }
+        let log_pow_col = ratio.ilog2() as usize;
```
Comment (adr1anh) [inline crates/ace-codegen/src/dag/lower.rs]: Log strict usize

File: crates/ace-codegen/src/dag/lower.rs [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+{
+    let mut acc = builder.constant(EF::ZERO);
+    for coeff in coeffs.iter().rev() {
+        let mul = builder.mul(point, acc);
+        acc = builder.add(*coeff, mul);
+    }
+    acc
+}
+
+fn input_key_for_symbolic<EF>(var: &SymVar<EF>) -> Result<InputKey, AceError> {
```
Comment (adr1anh) [inline crates/ace-codegen/src/dag/lower.rs]: Can't check in my editor right now, but should this be inlined into the lowering function?

File: crates/lib/core/tests/crypto/circuit_evaluation.rs [hunk 17]
```diff
@@ -81,3 +84,124 @@ fn circuit_evaluation_prove_verify() {
...
+    memory_felts.resize(padded_len, ZERO);
+    let num_adv_pipe = padded_len / 8;
+
+    let mut advice_builder = AdviceStackBuilder::new();
+    advice_builder.push_for_adv_pipe(&memory_felts);
+    let adv_stack = advice_builder.build_vec_u64();
+
+    // Place the circuit in memory at a fixed address.
+    let pointer = 1 << 16;
+    let num_read = encoded.num_vars();
```
Comment (huitseeker) [inline crates/lib/core/tests/crypto/circuit_evaluation.rs]: Bit confusing: If `eval_circuit` expects `read` rows, should this be `encoded.num_read_rows()`? If vars are correct, can we rename/comment to make the ABI expectation explicit?

File: crates/ace-codegen/src/encode.rs [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+
+impl<EF> AceCircuit<EF>
+where
+    EF: PrimeCharacteristicRing + BasedVectorSpace<Felt> + Copy + Eq + std::hash::Hash,
+{
+    /// Encode the circuit into the ACE chiplet format.
+    pub fn to_ace(&self) -> Result<EncodedCircuit, AceError> {
+        const MAX_NODE_ID: u64 = (1 << 30) - 1;
+
+        if EF::DIMENSION != EXPECTED_EXTENSION_DEGREE {
```
Comment (adr1anh) [inline crates/ace-codegen/src/encode.rs]: Could be a const assert

**Review by huitseeker — COMMENTED** [review]
It would be great to have a test for the `num_randomness == 2` edge case where `num_randomness_inputs == 2`, as well as direct‑randomness layouts (`num_randomness_inputs == num_randomness` with inputs supplied directly)

File: crates/ace-codegen/src/layout/mod.rs [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+//! extension-field elements in point-major order (all openings at `zeta`
+//! followed by all openings at `g * zeta`). Auxiliary trace and quotient
+//! chunk openings are provided as base-field coordinates and merged into
+//! extension elements inside the ACE circuit.
+//!
+//! All offsets and widths in this module are measured in extension-field (EF)
+//! elements. One EF element occupies two base-field elements.
+//!
+//! Besides the AIR constraints themselves, the ACE circuit computes:
+//! - Randomness expansion from `(alpha, beta)` into the full challenge vector `[alpha, 1, beta,
```
Comment (adr1anh) [inline crates/ace-codegen/src/layout/mod.rs]: Should we make this consistent with the challenges?

File: crates/ace-codegen/src/layout/plan.rs [hunk 20]
```diff
@@ -0,0 +1,228 @@
...
+    /// Use `num_randomness` for direct randomness, or `2` for alpha/beta seeds.
+    pub num_randomness_inputs: usize,
+    /// Number of periodic columns.
+    pub num_periodic: usize,
+    /// Number of auxiliary (stark var) inputs reserved.
+    pub num_aux_inputs: usize,
+    /// Number of quotient chunks.
+    pub num_quotient_chunks: usize,
+    /// Extension field degree.
+    pub ext_degree: usize,
```
Comment (adr1anh) [inline crates/ace-codegen/src/layout/plan.rs]: I think we should just keep ext degree as a constant everywhere. It requires extra checks, and would be a constant if we used Field/ExtensionField (I think)

File: crates/ace-codegen/src/layout/plan.rs [hunk 21]
```diff
@@ -0,0 +1,228 @@
...
+    pub aux_bus_boundary: InputRegion,
+    /// Stark variables (selectors, powers, weights).
+    pub stark_vars: InputRegion,
+}
+
+/// Indexes of canonical verifier scalars inside the stark-vars block.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(crate) struct StarkVarIndices {
+    /// Width of the base-field portion of the stark-vars block.
+    pub base_width: usize,
```
Comment (adr1anh) [inline crates/ace-codegen/src/layout/plan.rs]: Why is this necessary?

File: crates/ace-codegen/src/layout/policy.rs [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+    main: Alignment,
+    aux: Alignment,
+    quotient: Alignment,
+    aux_bus_boundary: Alignment,
+    stark_vars: Alignment,
+    end_align: Option<Alignment>,
+}
+
+impl LayoutPolicy {
+    fn native() -> Self {
```
Comment (adr1anh) [inline crates/ace-codegen/src/layout/policy.rs:25]: Do we need this?

File: crates/ace-codegen/src/pipeline.rs [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+/// Build a verifier-equivalent DAG and layout for the provided AIR.
+///
+/// This is useful when you need the DAG for off-VM checks and want to
+/// assemble inputs separately.
+pub(crate) fn build_ace_dag_for_air<A, F, EF>(
+    air: &A,
+    config: AceConfig,
+) -> Result<AceArtifacts<EF>, AceError>
```
Comment (adr1anh) [inline crates/ace-codegen/src/pipeline.rs:95]: I feel like we could probably simplify the whole thing to use assertions. The Air is supposed to be correct, and if not then everything else would break.

File: crates/ace-codegen/src/quotient.rs [hunk 24]
```diff
@@ -0,0 +1,238 @@
...
+    AceError,
+    dag::{DagBuilder, NodeId},
+    layout::{InputKey, InputLayout},
+};
+
+/// Evaluate the quotient recomposition at `zeta` using provided inputs.
+#[cfg(test)]
+pub fn eval_quotient<F, EF>(layout: &InputLayout, inputs: &[EF]) -> Result<EF, AceError>
+where
+    F: PrimeCharacteristicRing,
```
Comment (adr1anh) [inline crates/ace-codegen/src/quotient.rs]: Same as above

File: crates/ace-codegen/src/quotient.rs [hunk 25]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+/// Build DAG nodes that recombine quotient chunks.
+pub(crate) fn build_quotient_recomposition_dag<F, EF>(
+    builder: &mut DagBuilder<EF>,
+    layout: &InputLayout,
+) -> Result<NodeId, AceError>
+where
+    F: PrimeCharacteristicRing,
+    EF: PrimeCharacteristicRing + BasedVectorSpace<F> + Copy + Eq + Hash,
```
Comment (adr1anh) [inline crates/ace-codegen/src/quotient.rs]: Same as above

File: crates/ace-codegen/src/randomness.rs [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+    AceError,
+    dag::{DagBuilder, NodeId},
+    layout::{InputCounts, InputKey, InputLayout, InputRegion},
+};
+
+/// Randomness provisioning strategy for the ACE layout.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(crate) enum RandomnessPlan {
+    /// All randomness values are supplied directly.
+    Direct { num: usize },
```
Comment (adr1anh) [inline crates/ace-codegen/src/randomness.rs]: Do we still need this?

File: crates/ace-codegen/src/randomness.rs (hunk 10 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/randomness.rs]: Removed the `Direct` variant and simplified things now

File: crates/lib/core/tests/crypto/circuit_evaluation.rs (hunk 17 above)
Comment (Al-Kindi-0) [reply crates/lib/core/tests/crypto/circuit_evaluation.rs]: This is indeed confusing and I have fixed it, but we should think about unifying/improving the namings in other places

File: crates/ace-codegen/src/randomness.rs (hunk 12 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/randomness.rs]: Removed `Direct` and simplified

File: crates/ace-codegen/src/dag/ir.rs (hunk 1 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/dag/ir.rs:85]: done

File: crates/ace-codegen/src/dag/lower.rs (hunk 3 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/dag/lower.rs]: simplifed

File: crates/ace-codegen/src/dag/lower.rs (hunk 7 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/dag/lower.rs]: simplified

File: crates/ace-codegen/src/dag/lower.rs (hunk 15 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/dag/lower.rs]: Changed to using `u32`

File: crates/ace-codegen/src/dag/lower.rs (hunk 14 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/dag/lower.rs:135]: Indeed, fixed

File: crates/ace-codegen/src/dag/lower.rs (hunk 16 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/dag/lower.rs]: done

File: crates/ace-codegen/src/layout/policy.rs (hunk 22 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/layout/policy.rs:25]: This simplifies testing when we don't care about masm layout and is very little code overhead, so I think it is justified

File: crates/ace-codegen/src/layout/mod.rs (hunk 19 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/layout/mod.rs]: If I understand your comment correctly, I updated the comment and looked at re-using some of the infra for challenges we introduced. It seemed more work than it is worth it for now

File: crates/ace-codegen/src/quotient.rs (hunk 25 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/quotient.rs]: simplified

File: crates/ace-codegen/src/pipeline.rs (hunk 23 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/pipeline.rs:95]: Simplified

File: crates/ace-codegen/src/layout/plan.rs (hunk 21 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/layout/plan.rs]: Removed

File: crates/ace-codegen/src/randomness.rs (hunk 26 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/randomness.rs]: Removed `Direct` and then the whole enum now

File: crates/ace-codegen/src/dag/lower.rs (hunk 15 above)
Comment (adr1anh) [reply crates/ace-codegen/src/dag/lower.rs]: Nit, I meant using plonky3's log2_strict_usize, but tbh this is okay

File: crates/ace-codegen/src/layout/mod.rs (hunk 19 above)
Comment (adr1anh) [reply crates/ace-codegen/src/layout/mod.rs]: I wasn't sure, but don't we use powers of alpha instead of beta?

File: crates/ace-codegen/src/dag/builder.rs [hunk 27]
```diff
@@ -17,7 +17,7 @@ pub struct DagBuilder<EF> {
 
 impl<EF> DagBuilder<EF>
 where
-    EF: PrimeCharacteristicRing + Copy + Eq + std::hash::Hash,
+    EF: Field + Eq + std::hash::Hash,
```
Comment (adr1anh) [inline crates/ace-codegen/src/dag/builder.rs]: Just `Field` is enough

File: crates/ace-codegen/src/layout/mod.rs (hunk 19 above)
Comment (Al-Kindi-0) [reply crates/ace-codegen/src/layout/mod.rs]: we use powers of beta throughout.
Sometimes, we have `[beta, alpha]` but this is just a convenient way for MASM
